### PR TITLE
Refactor argparse options

### PR DESCRIPTION
**Pull request by Marshall-Hallenbeck**
Just a warning, a lot of this is just moving code around and renaming it, but in some of those instance its fixing bugs (see below about SMB opts).

- Allows us to pass in top level (universal) options ANYWHERE, meaning we can just append `--debug` to a command and it will work as intended.
- Fixes some other parsing issues, like some argparse groups in SMB were named the same (not sure how that didn't cause issues), and a lot of the groups had nonsense names.
- Adds a default option parser (`DisplayDefaultsNotNone`), so we do not have to include "Default: xyz" in the help information for options - it will automatically display those if they are not an empty string, None, False, etc.
- Updates commands into better groupings
- Adds testing the help for each protocol (might not need this, like the `--options` we just removed, @NeffIsBack?)

### CODE REVIEW

**Comment by Marshall-Hallenbeck** [comment]
Fixed conflicts with newly merged stuff (only big one was setting the retries to 10)

**Comment by Marshall-Hallenbeck** [comment]
Updated to include the new DNS options, so you can put them at the end now :)

**Review by mpgn — DISMISSED** [review]

File: nxc/protocols/smb/proto_args.py [hunk 1]
```diff
@@ -1,85 +1,87 @@
 from argparse import _StoreTrueAction
+from nxc.helpers.args import DisplayDefaultsNotNone
 
 
-def proto_args(parser, std_parser, module_parser):
-    smb_parser = parser.add_parser("smb", help="own stuff using SMB", parents=[std_parser, module_parser])
+def proto_args(parser, parents):
+    smb_parser = parser.add_parser("smb", help="own stuff using SMB", parents=parents, formatter_class=DisplayDefaultsNotNone)
     smb_parser.add_argument("-H", "--hash", metavar="HASH", dest="hash", nargs="+", default=[], help="NTLM hash(es) or file(s) containing NTLM hashes")
```
Comment (NeffIsBack) [inline nxc/protocols/smb/proto_args.py:7]: Imo this belongs to the "Authentication" section

File: nxc/protocols/smb/proto_args.py [hunk 2]
```diff
@@ -1,85 +1,87 @@
...
+    
     delegate_arg = smb_parser.add_argument("--delegate", action="store", help="Impersonate user with S4U2Self + S4U2Proxy")
     self_delegate_arg = smb_parser.add_argument("--self", dest="no_s4u2proxy", action=get_conditional_action(_StoreTrueAction), make_required=[], help="Only do S4U2Self, no S4U2Proxy (use with delegate)")
+    
     dgroup = smb_parser.add_mutually_exclusive_group()
-    dgroup.add_argument("-d", metavar="DOMAIN", dest="domain", type=str, help="domain to authenticate to")
+    dgroup.add_argument("-d", "--domain", metavar="DOMAIN", dest="domain", type=str, help="domain to authenticate to")
     dgroup.add_argument("--local-auth", action="store_true", help="authenticate locally to each target")
```
Comment (NeffIsBack) [inline nxc/protocols/smb/proto_args.py:14]: Same for this

File: nxc/protocols/smb/proto_args.py [hunk 3]
```diff
@@ -1,85 +1,87 @@
...
-    smb_parser.add_argument("--share", metavar="SHARE", default="C$", help="specify a share (default: %(default)s)")
+    
+    smb_parser.add_argument("--port", type=int, default=445, help="SMB port")
+    smb_parser.add_argument("--share", metavar="SHARE", default="C$", help="specify a share")
     smb_parser.add_argument("--smb-server-port", default="445", help="specify a server port for SMB", type=int)
     smb_parser.add_argument("--gen-relay-list", metavar="OUTPUT_FILE", help="outputs all hosts that don't require SMB signing to the specified file")
-    smb_parser.add_argument("--smb-timeout", help="SMB connection timeout, default %(default)s secondes", type=int, default=2)
+    smb_parser.add_argument("--smb-timeout", help="SMB connection timeout", type=int, default=2)
     smb_parser.add_argument("--laps", dest="laps", metavar="LAPS", type=str, help="LAPS authentification", nargs="?", const="administrator")
     self_delegate_arg.make_required = [delegate_arg]
```
Comment (NeffIsBack) [inline nxc/protocols/smb/proto_args.py:22]: Same for this

**Review by NeffIsBack — CHANGES_REQUESTED** [review]

**Comment by NeffIsBack** [comment]
Functionality looking good so far
image

**Review by NeffIsBack — APPROVED** [review]
LGTM, should be working as intended

**Review by NeffIsBack — APPROVED** [review]

**Pull request merged**